### PR TITLE
[Core] Support max cpu allocation per node for placement group scheduling

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

The PR adds a new experimental flag to the placement group API to avoid placement group taking all cpus on each node. It is used internally by Air to avoid placement group (created by Tune) is using all CPU resources which are needed for dataset

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scv119** [comment]
let's update the title and summary.

File: python/ray/util/placement_group.py [hunk 1]
```diff
@@ -147,6 +148,12 @@ def placement_group(
             will fate share with its creator and will be deleted once its
             creator is dead, or "detached", which means the placement group
             will live as a global object independent of the creator.
+        max_cpu_fraction_per_node: The maximum fraction of CPU cores this placement
```
Comment (scv119) [inline python/ray/util/placement_group.py:151]: let's keep this private?

Comment (scv119) [reply python/ray/util/placement_group.py:151]: also, as we discussed this is confusing if multiple placement group specify `max_cpu_fraction_per_node`. can we clarify in comment, or we rename the parameter to be more explicit.

Comment (rkooo567) [reply python/ray/util/placement_group.py:151]: done

Comment (ericl) [inline python/ray/util/placement_group.py:151]:
```suggestion
        _max_cpu_fraction_per_node: (Experimental) The maximum fraction of CPU cores this placement
```

File: python/ray/util/placement_group.py [hunk 2]
```diff
@@ -127,7 +127,8 @@ def placement_group(
     bundles: List[Dict[str, float]],
     strategy: str = "PACK",
     name: str = "",
-    lifetime=None,
+    lifetime: Optional[str] = None,
+    max_cpu_fraction_per_node: Optional[float] = None,
```
Comment (ericl) [inline python/ray/util/placement_group.py:131]:
```suggestion
    _max_cpu_fraction_per_node: Optional[float] = None,
```

File: python/ray/util/placement_group.py [hunk 3]
```diff
@@ -147,6 +148,12 @@ def placement_group(
             will fate share with its creator and will be deleted once its
             creator is dead, or "detached", which means the placement group
             will live as a global object independent of the creator.
+        max_cpu_fraction_per_node: The maximum fraction of CPU cores this placement
+            group can take up on each node. This must be a float between 0 and 1. If
+            provided, the placement group won't take up more than
+            max_cpu_fraction_per_node * node["num_cpus"] CPU cores on each node. This
+            is useful for ensuring that some percentage of CPU cores are available on
+            each node for workloads that aren't using this placement group.
```
Comment (ericl) [inline python/ray/util/placement_group.py:156]: Can we also document / guarantee that any non-zero value here will reserve at least 1 CPU per node?

**Review by scv119 — APPROVED** [review]
LGTM

**Comment by rkooo567** [comment]
ADdressed @scv119 ' comments, but I haven't pushed yet (finalizing unit tests).

**Comment by rkooo567** [comment]
I will merge it after pushing the code

File: python/ray/util/placement_group.py (hunk 3 above)
Comment (rkooo567) [reply python/ray/util/placement_group.py:156]: Do you think it's also reasonable to always round up? (e.g., 5 CPUs , 50% fraction -> can reserve up to 3 CPUs)

**Pull request merged**